sentence_transformers/trainer.py

```diff
@@ -0,0 +1,299 @@
+import logging
+import os
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union, TYPE_CHECKING
+
+import torch
+from torch import nn
+from torch.utils.data import DataLoader, ConcatDataset
+from transformers import PreTrainedTokenizerBase, Trainer, EvalPrediction, TrainerCallback
+from transformers.trainer import TRAINING_ARGS_NAME
+from transformers.training_args import ParallelMode
+
+from datasets import DatasetDict, Dataset
+from transformers.trainer_utils import EvalLoopOutput
+from transformers.data.data_collator import DataCollator
+from sentence_transformers.losses import CoSENTLoss
+
+from sentence_transformers.training_args import TrainingArguments
+from sentence_transformers.data_collator import SentenceTransformerDataCollator
+from sentence_transformers.evaluation import SentenceEvaluator
+from sentence_transformers.sampler import ProportionalBatchSampler, RoundRobinBatchSampler
+
+logger = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from sentence_transformers.SentenceTransformer import SentenceTransformer
+
+
+class SentenceTransformerTrainer(Trainer):
+    def __init__(
+        self,
+        model: Optional["SentenceTransformer"] = None,
```

Comment:
not sure about this type hints, when we want to train a model into embedding model, we have 2 cases:
1. train from a foundation model, such as BERT or T5 like.
2. fine-tune an existent sentence-transformer model. Then`Optional["SentenceTransofmer"]` make sense.

besides, why it is an optional parameter?

Reply:
When you train from a foundation model, such as BERT, T5-like, MPNet, etc., then the recommendation is to create a new `SentenceTransformer` model using these checkpoints. Either use the default option (just the model + mean pooling):
```python
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("bert-base-cased")
"""
No sentence-transformers model found with name bert-base-cased. Creating a new one with MEAN pooling.
SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
)
"""
# or
model = SentenceTransformer("bert-base-cased")
"""
No sentence-transformers model found with name google-t5/t5-base. Creating a new one with MEAN pooling.
SentenceTransformer(
  (0): Transformer({'max_seq_length': None, 'do_lower_case': False}) with Transformer model: T5EncoderModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
)
"""
# or
model = SentenceTransformer("microsoft/mpnet-base")
"""
No sentence-transformers model found with name microsoft/mpnet-base. Creating a new one with MEAN pooling.
Some weights of MPNetModel were not initialized from the model checkpoint at microsoft/mpnet-base and are newly initialized: ['mpnet.pooler.dense.bias', 'mpnet.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: MPNetModel
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
)
"""
```

Or you can be more specific by using ST modules:
```python
from sentence_transformers.models import Transformer, Pooling, Normalize
from sentence_transformers import SentenceTransformer

transformer = Transformer("google-t5/t5-base", max_seq_length=384)
pooling = Pooling(transformer.get_word_embedding_dimension())
normalize = Normalize()
model = SentenceTransformer(modules=[transformer, pooling, normalize])
"""
SentenceTransformer(
  (0): Transformer({'max_seq_length': 384, 'do_lower_case': False}) with Transformer model: T5EncoderModel
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)
"""
```

If you are looking to first continue pretraining the foundational model using e.g. Masked Language Modeling, then I would advise to use pure `transformers`, and then only use Sentence Transformers whenever you start finetuning with e.g. sentence pairs/triplets.

Does that make sense? So, this way you can always get a `SentenceTransformer` instance from your foundational model or an already finetuned model.

Reply:
Also, it can be optional because you must either provide a `model` (99% of cases) or a `model_init`, a function that returns a model. This is only really useful when you are doing some form of hyperparameter optimization. See also the hyperparameter tuning documentation in Transformers [here](https://huggingface.co/docs/transformers/en/hpo_train).